src/pudl/analysis/record_linkage/cross_year.py

```diff
@@ -0,0 +1,241 @@
+                        n_clusters=self.n_clusters,
+                        metric="precomputed",
+                        linkage="average",
+                        distance_threshold=self.distance_threshold,
+                        compute_distances=True,
+                    ),
+                ),
+            ]
+        )
+
+        df[self.id_column] = pipeline.fit_predict(df)
+
+        return df
+
+
+class PrecomputeDistance(BaseEstimator, TransformerMixin):
+    """Precompute distances which are used during clustering and prediction."""
+
+    def __init__(self, metric="euclidean"):
+        """Initialize with distance metric."""
+        self.metric = metric
+
+    def fit(self, X, y=None, **fit_params):  # noqa: N803
+        """Required by scikit-learn, but does not modify anything."""
+        return self
+
+    def transform(self, X, y=None, **fit_params):  # noqa: N803
+        """Compute distance between records then add penalty to records from same year."""
+        return pairwise_distances(X, metric=self.metric)
+
+
+class DistancePenalizeSameYear(PrecomputeDistance):
+    """Custom estimator to compute distances used to identify clusters of plants."""
+
+    def __init__(self, report_years: np.array, metric="euclidean", penalty=100):
+        """Initialize estimator with configurable parameters.
+
+        Args:
+            report_years: reporty_year column used to penalize distance for records
+                          from same year.
+            metric: Distance metric to use in computation.
+            penalty: Penalty to apply to records with the same report year.
+        """
+        self.report_years = report_years
+        self.metric = metric
+        self.penalty = penalty
+
+    def fit(self, X, y=None, **fit_params):  # noqa: N803
+        """Required by scikit-learn, but does not modify anything."""
+        return self
+
+    def transform(self, X, y=None, **fit_params):  # noqa: N803
+        """Compute distance between records then add penalty to records from same year."""
+        dist_matrix = super().transform(X)
+
+        # Create penalty matrix
+        # Probably not the most elegant way to handle this
+        penalty_matrix = pairwise_distances(self.report_years.reshape(-1, 1))
+        penalty_matrix += self.penalty
+        penalty_matrix[penalty_matrix > self.penalty] = 0
+
+        # distance from node to itself should still be 0
+        np.fill_diagonal(penalty_matrix, 0)
+        dist_matrix += penalty_matrix
+        return dist_matrix
+
+
+class DenseTransformer(BaseEstimator, TransformerMixin):
+    """Convert sparse numpy matrix to dense numpy array."""
+
+    def fit(self, X, y=None, **fit_params):  # noqa: N803
+        """No modifications made during fitting."""
+        return self
+
+    def transform(self, X, y=None, **fit_params):  # noqa: N803
+        """No modifications made during fitting."""
+        return np.asarray(np.float32(X.todense()))
```

Comment:
Now I'm not remembering if converting to `np.float32` actually matters. I think that was only required if the dense matrix was going to be an input to an `faiss` index. I guess it's fine to keep as float32?